<a href="https://colab.research.google.com/github/unt-iialab/INFO5731_Spring2020/blob/master/Assignments/INFO5731_Assignment_One.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment - 1**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.


1.  Question 1-5 using AI assistance is strictly prohibted
2.  Question 6-7 AI assitance is recommended by following guidelines.

Guidelines and expectations file can be found in week 2 module.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
x = 2/3 #assigned x value as 2/3 according to formulae
p = 22/7 #assigned pi value  to y

print("r", "V")
for i in range(1, 21):  #using for loop in range of 1 to 21
  if i%2 == 0:          #verfying the integer is even, if even integer goes to next step
    r = i**3            #calaculating the cube of radius
    V = x*p*r           # multilplying the all values to get Volume
    print(i, V)


r V
2 16.76190476190476
4 134.09523809523807
6 452.5714285714285
8 1072.7619047619046
10 2095.238095238095
12 3620.571428571428
14 5749.333333333332
16 8582.095238095237
18 12219.428571428569
20 16761.90476190476


# Question 2 (10 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [2]:
# Write code for each of the sub parts with proper comments.
import numpy as np                                              #imported numpy library to generate random numbers

x = np.random.randint(50, 151, size=12)                        #By using random in range of 50 to 151 with size of 12 generated 12 random numbers
print("Sequence random numbers:",x)

x.sort()                                                       #By using sort() function sorted the random numbers in permanently in ascending order
print("Ascending Order:",x)

y = x[::-1]                                                    #By slicing list from last to first with negative index to get descending order
print("Descending Order:",y)

av = sum(x)/12
print("Avergae of entier list is :", av)                        #calculated the average of list by sum function and total size of list

fisrt6_asc_avg = sum(x[6:])/6                                   #calculated the average of first 6 numbers of ascending list by slicing the list in sum function
print("Average of Fisrt 6 ascending numbers:",fisrt6_asc_avg)

first6_dsc_avg = sum(y[6:])/6                                   #calculated the average of first 6 numbers of descending list by slicing the list in sum function
print("Average of Fisrt 6 descending numbers:",first6_dsc_avg)

Sequence random numbers: [106 117 119 117 127  88  73 143  58  78  66 101]
Ascending Order: [ 58  66  73  78  88 101 106 117 117 119 127 143]
Descending Order: [143 127 119 117 117 106 101  88  78  73  66  58]
Avergae of entier list is : 99.41666666666667
Average of Fisrt 6 ascending numbers: 121.5
Average of Fisrt 6 descending numbers: 77.33333333333333


# Question 3 (10 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [21]:
import string


#Part-A
input1 = input("Enter the sentence1: ")  #inputing the 1st sentence
input2 = input("Enter the sentence2: ")  #inputing the 2nd sentence

#Part-B removing punctuation

a = input1.split()  # spilting the each word of input 1
b = input2.split()  # spilting the each word in input2 sentence

sentence1 = [] # intiating the empty sentence lists two add each word of sentence
sentence2 = []
pun = string.punctuation   # intaiting the pun to all puncations stoer in string.puncations method

for i in a:   #using for loop

    if i[-1] in pun:   # verfying the each word in puncations in neagtive index
      sentence1.append(i[:-1])  # stroing the puncations in sentence1 list
      sentence1.append(i[-1])
    else:
      sentence1.append(i) # if word doesnt have the puncation directly it will append to sentence 1 list

for i in b:     # same as above loop

    if i[-1] in pun:
      sentence2.append(i[:-1])
      sentence2.append(i[-1])
    else:
      sentence2.append(i)
print("Sanitize Sentences: ")
print(sentence1) # print the result
print(sentence2)

#Part-C Compare the simsilar words in sentences

compare_list = []  # intating the empty list

for i in sentence1:   # in this loop coparing the each word in sentence,  right now taking the 1st word in sentence1[0] list
  for j in sentence2: # in this innner loop taking the 1st word in sentece2[0] loop,  after completing the sentece2[0] it will the verfy upto sentence2[n](last index word in sentence2) after completing the comlete len of sentnece 2 it verfys the sentence[1]
    if i == j:   # comparing the sentence1[0] and senrecne[0] words
      compare_list.append(i) # if it finds similar  append to compare list

print("Similar Words: ",compare_list)


#Part-D Remove ,:;.

santized_sentence1 = ''  # intaiting the new sentence1 string
santized_sentence2 = ''  # intaiting the new sentence2 string

for i in input1:  # using for loop, taking the input1 first character
  if i != '.' and i != ',' and i != ';' and i != ':':   # verfying  char is not falls in conditions
    santized_sentence1 += i   # if char not falls in condition it  update into santized_sentence1 string

for i in input2:
  if i != '.' and i != ',' and i != ';' and i != ':':
    santized_sentence2 += i


print(santized_sentence1)
print(santized_sentence2)


#Part-E couunt the vowles
input1 = input1.lower()  #converting the input1 sentence into lower case for dictionary to count the vowels
input2 = input2.lower()  #converting the input2 sentence into lower case for dictionary to count the vowels

vowels = 'aeiou' # intiating vowels to vowels

dict1 = {'a': 0, 'e': 0, 'i': 0, 'o': 0, 'u': 0} # intiaiting the dictonary(dict1) with null count of each vowel
dict2 = {'a': 0, 'e': 0, 'i': 0, 'o': 0, 'u': 0} # intiaiting the dictonary(dict2) with null count of each vowel

for i in input1:  # using for loop, to count vowels veryfing the each char input1(0)....input(n)
  if i in vowels:  # verfying the 1st char of falls in vowles
    dict1[i] += 1  # if char in vowles it will upadte the count into dict1

for i in input2:
  if i in vowels:
    dict2[i] += 1

print("Sentenc1: ")
for i , j in dict1.items():   #using the for loop, pirnting the each item in dict as i and j
  if j > 0: # if any count more than zero results are printing
    print(i, ":", j)

print("Sentenc2: ")
for i , j in dict2.items():
  if j > 0:
    print(i, ":", j)

#Hello, how are you doing today?
#This is an example sentence for the program!

Enter the sentence1: Hello, how are you doing today?
Enter the sentence2: This is an example sentence for the program!
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?']
['This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']
Similar Words:  []
Hello how are you doing today?
This is an example sentence for the program!
Sentenc1: 
a : 2
e : 2
i : 1
o : 5
u : 1
Sentenc2: 
a : 3
e : 6
i : 2
o : 2


# Question 4 (10 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [102]:
import math
import re

def calculator(E): # calculator Function is used to take math expresion and evaluates it and returns the value
    try:
        E = E.replace('^', '**')


        # Function to evaluate expressions with parentheses
        def evaluate(tokens):
            def apply_operator(operators, values):
                right = values.pop()
                left = values.pop() if values else 0  # Handle unary operations
                op = operators.pop()
                if op == '+':
                    values.append(left + right)
                elif op == '-':
                    values.append(left - right)
                elif op == '*':
                    values.append(left * right)
                elif op == '/':
                    if right == 0:
                        raise ZeroDivisionError("Division by zero is not allowed.")
                    values.append(left // right)  # Integer division
                elif op == '**':
                    values.append(left ** right)
                elif op == '%':
                    values.append(left % right)

            # Define operator precedence
            precedence = {'+': 1, '-': 1, '': 2, '/': 2, '*': 3, '%': 2}
            values = [] # Stack for numbers
            operators = [] # Stack for operator
            i = 0 # Token index

            while i < len(tokens):
                token = tokens[i]
                # Handle numbers
                if isinstance(token, str) and token.isdigit():
                    values.append(int(token))
                elif token == 'sqrt': # Handle square root function
                    if i + 1 < len(tokens) and tokens[i + 1] == '(':
                        operators.append(token)
                    else:
                        return "Error: sqrt function requires parentheses."
                elif token in precedence: # Handle arithmetic operators
                    while (operators and operators[-1] in precedence and
                           precedence[operators[-1]] >= precedence[token]):
                        apply_operator(operators, values)
                    operators.append(token)
                elif token == '(':
                    operators.append(token)
                elif token == ')':
                    while operators and operators[-1] != '(':
                        apply_operator(operators, values)
                    operators.pop()  # Remove '('
                    if operators and operators[-1] == 'sqrt':
                        operators.pop()
                        values.append(int(math.sqrt(values.pop())))
                i += 1

            while operators:
                apply_operator(operators, values)

            return values[0] if values else "Error: Invalid expression."

        tokens = re.findall(r'sqrt|\d+|[()+\-*/%^]', E) # Tokenize the input expression
        return evaluate(tokens)
    except ZeroDivisionError:
        return "Error: Dividing by Zero is not possible. Try again"
    except (SyntaxError, NameError, TypeError):
        return "Error: Thats not a valid input. Please enter a valid mathematical expression."

def main():
    print("Hello!! Welcome to karthik's Command line Calculator!")
    print("Valid Operations: +, -, *, /, ^, %, sqrt()")
    print("type 'exit' to quit the calculator.")

    while True: # looping for user input
        userInput = input("Enter a math expression:")
        if userInput.lower() == 'exit':
            print("Thank you for using my calculator :)")
            break
        output = calculator(userInput)
        print("Result:", output)


if __name__ == "__main__":
    main()


Hello!! Welcome to karthik's Command line Calculator!
Valid Operations: +, -, *, /, ^, %, sqrt()
type 'exit' to quit the calculator.
Enter a math expression:exit
Thank you for using my calculator :)


# Question 5 (20 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (5 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (5 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [27]:
import numpy as np
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv")

print(df)

    year             title_field  value
0   2014     Total Housing Units  49109
1   2014    Vacant Housing Units   2814
2   2014  Occupied Housing Units  46295
3   2013     Total Housing Units  47888
4   2013    Vacant Housing Units   4215
5   2013  Occupied Housing Units  43673
6   2012     Total Housing Units  45121
7   2012    Vacant Housing Units   3013
8   2012  Occupied Housing Units  42108
9   2011     Total Housing Units  44917
10  2011    Vacant Housing Units   4213
11  2011  Occupied Housing Units  40704
12  2010     Total Housing Units  44642
13  2010    Vacant Housing Units   3635
14  2010  Occupied Housing Units  41007
15  2009     Total Housing Units  39499
16  2009    Vacant Housing Units   3583
17  2009  Occupied Housing Units  35916
18  2008     Total Housing Units  41194
19  2008    Vacant Housing Units   4483
20  2008  Occupied Housing Units  36711
21  2014          Owner Occupied  21427
22  2014         Renter Occupied  24868
23  2013          Owner Occupied  20845


In [60]:
# Your code here
a = df[df['title_field'] == 'Occupied Housing Units'] # filtered out the only Occupied Housing Units data
a = a.sort_values(by='year') # sorted by year the data
a['difference'] = a['value'].diff()  # calculating the consecytive difference
a = a.dropna(subset=['difference']) # droping the null values in data

list = [] # intiatiezed the empty new list


for i in range(len(a) - 1):  # using for loop, from a data of a[0](year1, year2, diiference value) appending to new list
    y1 = a.iloc[i]['year']
    y2 = a.iloc[i + 1]['year']
    res = a.iloc[i + 1]['difference']
    list.append((y1, y2, res))


x = sorted(list, key=lambda x: x[2]) # sorting the new list in ascending order by difference values in X

for y1, y2, diff in x: # printing the result in given formart as 2009-2010 and difference
    print(f"{y1} - {y2}: {diff}")


2010 - 2011: -303.0
2011 - 2012: 1404.0
2012 - 2013: 1565.0
2013 - 2014: 2622.0
2009 - 2010: 5091.0


In [70]:
a = df[df['title_field'] == 'Vacant Housing Units'] # storing only the Vacant Housing Units data in a
b = df[df['title_field'] == 'Occupied Housing Units']  # storing only the Occupied Housing Units data in b
c = df[df['title_field'] == 'Total Housing Units']  # storing only the Occupied Housing Units data in c


res_df = a[['year', 'value']].merge(b[['year', 'value']], on='year', suffixes=('_vacant', '_occupied')) # mergeing the a nd b of Vacant Housing Units and Occupied Housing Units data
res_df = res_df.merge(c[['year', 'value']], on='year', suffixes=('', '_total'))  # merging the above merged data and Occupied Housing Units data

res_df['percentage_vacant'] = (res_df['value_vacant'] / res_df['value']) * 100. #calaculating the percentgae of Vacant Housing Units and total housing
res_df['percentage_occupied'] = (res_df['value_occupied'] / res_df['value']) * 100. #calaculating the percentgae of ccupied Housing Units and total housing


for _, row in res_df.iterrows():   #printing the output in given format
    print(f"{row['year']}  {row['percentage_vacant']:.2f}  {row['percentage_occupied']:.2f}")


2014.0  5.73  94.27
2013.0  8.80  91.20
2012.0  6.68  93.32
2011.0  9.38  90.62
2010.0  8.14  91.86
2009.0  9.07  90.93
2008.0  10.88  89.12


In [91]:
a = df[df['title_field'] == 'Vacant Housing Units'] # storing only the Vacant Housing Units data in a
b = df[df['title_field'] == 'Occupied Housing Units']  # storing only the Occupied Housing Units data in b

a = a.loc[a['value'].idxmax()] # using idmax method to get the maximum value of Vacant Housing Units
b = b.loc[b['value'].idxmax()] # using idmax methos to get the maximum values of Occupied Housing Units

# Print the results in the desired format
print(f"                       Year Value ")
print(f"Vacant Housing Units   {a['year']} {a['value']} ")
print(f"Occupied Housing Units {b['year']} {b['value']} ")

                       Year Value 
Vacant Housing Units   2008 4483 
Occupied Housing Units 2014 46295 


In [94]:
c = df[df['title_field'] == 'Total Housing Units']  # storing only the Occupied Housing Units data in c
values = c['value'] # storing the only totoal housing units of values into values
i = len(values) #calculating the len of values
hm = i / (1 / values).sum() # harmonic formaulae len(values)/(1/ sum of values)
print('Harmonic Mean of Total Housing Units: ',hm)


Harmonic Mean of Total Housing Units:  44398.28543969185


# **For question 6 and 7 you can take AI assitance to solve.**

Make sure you follow the instructions given in class to answer these questions.


**Important points:**

1.   You must come up with at least 10 edge cases.
2.   You should provide prompts and revised prompts given to ChatGPT.
2.   Debug your code and write comments for each block of code
3.   Analyzing error will not consider as edge case.

Edge cases are like real life scienarios which are realated to the given question.

# Question 6 (20 points).
Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

#1
write a python program for smart inventory managing to product managing company. products with id, name, serial number, category, price, date added
*italicized text*

In [110]:
import datetime

class Product:
    def __init__(self, product_id, name, serial_number, category, price):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = datetime.datetime.now()

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def update_product(self, product_id, name=None, serial_number=None, category=None, price=None):
        if product_id in self.products:
            product = self.products[product_id]
            if name:
                product.name = name
            if serial_number:
                product.serial_number = serial_number
            if category:
                product.category = category
            if price is not None:
                product.price = price
            print(f"Product ID {product_id} updated successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def delete_product(self, product_id):
        if product_id in self.products:
            del self.products[product_id]
            print(f"Product ID {product_id} deleted successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def view_product(self, product_id):
        if product_id in self.products:
            print(self.products[product_id])
        else:
            print(f"Product with ID {product_id} not found.")

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

def main():
    inventory = Inventory()

    # Menu for user interaction
    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Update Product")
        print("3. Delete Product")
        print("4. View Product")
        print("5. View All Products")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            # Add Product
            product_id = input("Enter product ID: ")
            name = input("Enter product name: ")
            serial_number = input("Enter serial number: ")
            category = input("Enter product category: ")
            price = float(input("Enter product price: "))

            product = Product(product_id, name, serial_number, category, price)
            inventory.add_product(product)

        elif choice == '2':
            # Update Product
            product_id = input("Enter product ID to update: ")
            print("Leave fields blank to keep current values.")
            name = input("Enter new name (or leave blank): ")
            serial_number = input("Enter new serial number (or leave blank): ")
            category = input("Enter new category (or leave blank): ")
            price_input = input("Enter new price (or leave blank): ")
            price = float(price_input) if price_input else None

            inventory.update_product(product_id, name, serial_number, category, price)

        elif choice == '3':
            # Delete Product
            product_id = input("Enter product ID to delete: ")
            inventory.delete_product(product_id)

        elif choice == '4':
            # View Product
            product_id = input("Enter product ID to view: ")
            inventory.view_product(product_id)

        elif choice == '5':
            # View All Products
            inventory.view_all_products()

        elif choice == '6':
            # Exit
            print("Exiting the system...")
            break

        else:
            print("Invalid choice, please try again.")

if __name__ == "__main__":
    main()



Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 11
Invalid choice, please try again.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 1
Enter product ID: 11
Enter product name: 1
Enter serial number: 334
Enter product category: frf
Enter product price: 34
Product 1 added successfully.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit


KeyboardInterrupt: Interrupted by user

#2
In above code enciunterd some issues while adding the product
rewrite the code by following instructions of product id should be integer within  length of 10, product name should contain at least of 3 chars(doesn't specify any length), serial number should be as combination of string and integer(doesn't specify any length).





In [2]:
import datetime
import re

class Product:
    def __init__(self, product_id, name, serial_number, category, price):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = datetime.datetime.now()

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def update_product(self, product_id, name=None, serial_number=None, category=None, price=None):
        if product_id in self.products:
            product = self.products[product_id]
            if name:
                product.name = name
            if serial_number:
                product.serial_number = serial_number
            if category:
                product.category = category
            if price is not None:
                product.price = price
            print(f"Product ID {product_id} updated successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def delete_product(self, product_id):
        if product_id in self.products:
            del self.products[product_id]
            print(f"Product ID {product_id} deleted successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def view_product(self, product_id):
        if product_id in self.products:
            print(self.products[product_id])
        else:
            print(f"Product with ID {product_id} not found.")

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

    def validate_product_id(self, product_id):
        """Validate that product ID is a 10-digit integer."""
        return isinstance(product_id, int) and len(str(product_id)) == 10

    def validate_product_name(self, name):
        """Validate that the product name has at least 3 characters."""
        return len(name) >= 3

    def validate_serial_number(self, serial_number):
        """Validate serial number is alphanumeric with a combination of letters and digits."""
        return bool(re.match(r'^[A-Za-z]+\d+$', serial_number))

def main():
    inventory = Inventory()

    # Menu for user interaction
    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Update Product")
        print("3. Delete Product")
        print("4. View Product")
        print("5. View All Products")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            # Add Product
            try:
                product_id = int(input("Enter product ID (exactly 10 digits): "))
                if not inventory.validate_product_id(product_id):
                    print("Error: Product ID must be exactly 10 digits.")
                    continue
            except ValueError:
                print("Error: Product ID must be an integer.")
                continue

            name = input("Enter product name (at least 3 characters): ")
            if not inventory.validate_product_name(name):
                print("Error: Product name must be at least 3 characters.")
                continue

            serial_number = input("Enter serial number (letters followed by numbers, e.g., ABC1234567): ")
            if not inventory.validate_serial_number(serial_number):
                print("Error: Serial number must be a combination of letters and numbers (e.g., ABC1234567).")
                continue

            category = input("Enter product category: ")
            try:
                price = float(input("Enter product price: "))
            except ValueError:
                print("Error: Price must be a valid number.")
                continue

            product = Product(product_id, name, serial_number, category, price)
            inventory.add_product(product)

        elif choice == '2':
            # Update Product
            product_id = input("Enter product ID to update: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)

            print("Leave fields blank to keep current values.")
            name = input("Enter new name (or leave blank): ")
            serial_number = input("Enter new serial number (or leave blank): ")
            category = input("Enter new category (or leave blank): ")
            price_input = input("Enter new price (or leave blank): ")
            price = float(price_input) if price_input else None

            inventory.update_product(product_id, name, serial_number, category, price)

        elif choice == '3':
            # Delete Product
            product_id = input("Enter product ID to delete: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.delete_product(product_id)

        elif choice == '4':
            # View Product
            product_id = input("Enter product ID to view: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.view_product(product_id)

        elif choice == '5':
            # View All Products
            inventory.view_all_products()

        elif choice == '6':
            # Exit
            print("Exiting the system...")
            break

        else:
            print("Invalid choice, please try again.")

if __name__ == "__main__":
    main()



Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 1
Enter product ID (exactly 10 digits): 1234567890
Enter product name (at least 3 characters): gerg
Enter serial number (letters followed by numbers, e.g., ABC1234567): adc234
Enter product category: dcee
Enter product price: 34
Product gerg added successfully.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 1
Enter product ID (exactly 10 digits): ef
Error: Product ID must be an integer.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: exit 
Invalid choice, please try again.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: Exit
Invalid choice, please try again.

I

#3,  Modify the code according to: if mistake is during the entering the product data entry details, should ask the same question, don't skip to main menu


In [3]:
import datetime
import re

class Product:
    def __init__(self, product_id, name, serial_number, category, price):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = datetime.datetime.now()

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def update_product(self, product_id, name=None, serial_number=None, category=None, price=None):
        if product_id in self.products:
            product = self.products[product_id]
            if name:
                product.name = name
            if serial_number:
                product.serial_number = serial_number
            if category:
                product.category = category
            if price is not None:
                product.price = price
            print(f"Product ID {product_id} updated successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def delete_product(self, product_id):
        if product_id in self.products:
            del self.products[product_id]
            print(f"Product ID {product_id} deleted successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def view_product(self, product_id):
        if product_id in self.products:
            print(self.products[product_id])
        else:
            print(f"Product with ID {product_id} not found.")

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

    def validate_product_id(self, product_id):
        """Validate that product ID is a 10-digit integer."""
        return isinstance(product_id, int) and len(str(product_id)) == 10

    def validate_product_name(self, name):
        """Validate that the product name has at least 3 characters."""
        return len(name) >= 3

    def validate_serial_number(self, serial_number):
        """Validate serial number is alphanumeric with a combination of letters and digits."""
        return bool(re.match(r'^[A-Za-z]+\d+$', serial_number))

def get_valid_input(prompt, validation_func, error_message):
    """Helper function to get valid input from the user with custom validation."""
    while True:
        user_input = input(prompt)
        if validation_func(user_input):
            return user_input
        else:
            print(error_message)

def main():
    inventory = Inventory()

    # Menu for user interaction
    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Update Product")
        print("3. Delete Product")
        print("4. View Product")
        print("5. View All Products")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            # Add Product with error handling
            print("\n--- Add Product ---")
            try:
                product_id = get_valid_input(
                    "Enter product ID (exactly 10 digits): ",
                    lambda x: x.isdigit() and len(x) == 10,
                    "Error: Product ID must be exactly 10 digits."
                )
                product_id = int(product_id)
            except ValueError:
                print("Error: Product ID must be an integer.")
                continue

            name = get_valid_input(
                "Enter product name (at least 3 characters): ",
                lambda x: len(x) >= 3,
                "Error: Product name must be at least 3 characters."
            )

            serial_number = get_valid_input(
                "Enter serial number (letters followed by numbers, e.g., ABC1234567): ",
                lambda x: bool(re.match(r'^[A-Za-z]+\d+$', x)),
                "Error: Serial number must be a combination of letters and numbers (e.g., ABC1234567)."
            )

            category = input("Enter product category: ")
            try:
                price = float(input("Enter product price: "))
            except ValueError:
                print("Error: Price must be a valid number.")
                continue

            product = Product(product_id, name, serial_number, category, price)
            inventory.add_product(product)

        elif choice == '2':
            # Update Product
            product_id = input("Enter product ID to update: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)

            print("Leave fields blank to keep current values.")
            name = input("Enter new name (or leave blank): ")
            serial_number = input("Enter new serial number (or leave blank): ")
            category = input("Enter new category (or leave blank): ")
            price_input = input("Enter new price (or leave blank): ")
            price = float(price_input) if price_input else None

            inventory.update_product(product_id, name, serial_number, category, price)

        elif choice == '3':
            # Delete Product
            product_id = input("Enter product ID to delete: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.delete_product(product_id)

        elif choice == '4':
            # View Product
            product_id = input("Enter product ID to view: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.view_product(product_id)

        elif choice == '5':
            # View All Products
            inventory.view_all_products()

        elif choice == '6':
            # Exit
            print("Exiting the system...")
            break

        else:
            print("Invalid choice, please try again.")

if __name__ == "__main__":
    main()



Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 1

--- Add Product ---
Enter product ID (exactly 10 digits): 1234567890
Enter product name (at least 3 characters): efer
Enter serial number (letters followed by numbers, e.g., ABC1234567): asdc34
Enter product category: derf
Enter product price: 234
Product efer added successfully.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 1

--- Add Product ---
Enter product ID (exactly 10 digits): 234
Error: Product ID must be exactly 10 digits.
Enter product ID (exactly 10 digits): 333
Error: Product ID must be exactly 10 digits.
Enter product ID (exactly 10 digits): 3333333333
Enter product name (at least 3 characters): fhfj
Enter serial number (letters followed by numbers, e.g., ABC1234567): sd34
Enter product category: ffefrg
Enter product price: 45
Prod

#4 Rewrite the code following instructions as: In Update product option only, give the what kind of options need to update, and take the input following selected option


In [3]:
import datetime
import re

class Product:
    def __init__(self, product_id, name, serial_number, category, price):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = datetime.datetime.now()

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def update_product(self, product_id):
        if product_id in self.products:
            product = self.products[product_id]
            print("\nSelect the attribute to update:")
            print("1. Update Name")
            print("2. Update Serial Number")
            print("3. Update Category")
            print("4. Update Price")
            choice = input("Enter your choice (1-4): ")

            if choice == '1':
                new_name = input(f"Enter new name for product (current: {product.name}): ")
                if len(new_name) >= 3:
                    product.name = new_name
                    print(f"Product name updated to: {new_name}")
                else:
                    print("Error: Product name must be at least 3 characters.")
            elif choice == '2':
                new_serial_number = input(f"Enter new serial number for product (current: {product.serial_number}): ")
                if bool(re.match(r'^[A-Za-z]+\d+$', new_serial_number)):
                    product.serial_number = new_serial_number
                    print(f"Product serial number updated to: {new_serial_number}")
                else:
                    print("Error: Serial number must be a combination of letters and numbers.")
            elif choice == '3':
                new_category = input(f"Enter new category for product (current: {product.category}): ")
                product.category = new_category
                print(f"Product category updated to: {new_category}")
            elif choice == '4':
                try:
                    new_price = float(input(f"Enter new price for product (current: ${product.price}): "))
                    product.price = new_price
                    print(f"Product price updated to: ${new_price:.2f}")
                except ValueError:
                    print("Error: Price must be a valid number.")
            else:
                print("Invalid choice! No update was made.")
        else:
            print(f"Product with ID {product_id} not found.")

    def delete_product(self, product_id):
        if product_id in self.products:
            del self.products[product_id]
            print(f"Product ID {product_id} deleted successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def view_product(self, product_id):
        if product_id in self.products:
            print(self.products[product_id])
        else:
            print(f"Product with ID {product_id} not found.")

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

    def validate_product_id(self, product_id):
        """Validate that product ID is a 10-digit integer."""
        return isinstance(product_id, int) and len(str(product_id)) == 10

    def validate_product_name(self, name):
        """Validate that the product name has at least 3 characters."""
        return len(name) >= 3

    def validate_serial_number(self, serial_number):
        """Validate serial number is alphanumeric with a combination of letters and digits."""
        return bool(re.match(r'^[A-Za-z]+\d+$', serial_number))

def get_valid_input(prompt, validation_func, error_message):
    """Helper function to get valid input from the user with custom validation."""
    while True:
        user_input = input(prompt)
        if validation_func(user_input):
            return user_input
        else:
            print(error_message)

def main():
    inventory = Inventory()

    # Menu for user interaction
    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Update Product")
        print("3. Delete Product")
        print("4. View Product")
        print("5. View All Products")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            # Add Product with error handling
            print("\n--- Add Product ---")
            try:
                product_id = get_valid_input(
                    "Enter product ID (exactly 10 digits): ",
                    lambda x: x.isdigit() and len(x) == 10,
                    "Error: Product ID must be exactly 10 digits."
                )
                product_id = int(product_id)
            except ValueError:
                print("Error: Product ID must be an integer.")
                continue

            name = get_valid_input(
                "Enter product name (at least 3 characters): ",
                lambda x: len(x) >= 3,
                "Error: Product name must be at least 3 characters."
            )

            serial_number = get_valid_input(
                "Enter serial number (letters followed by numbers, e.g., ABC1234567): ",
                lambda x: bool(re.match(r'^[A-Za-z]+\d+$', x)),
                "Error: Serial number must be a combination of letters and numbers (e.g., ABC1234567)."
            )

            category = input("Enter product category: ")
            try:
                price = float(input("Enter product price: "))
            except ValueError:
                print("Error: Price must be a valid number.")
                continue

            product = Product(product_id, name, serial_number, category, price)
            inventory.add_product(product)

        elif choice == '2':
            # Update Product
            product_id = input("Enter product ID to update: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)

            inventory.update_product(product_id)

        elif choice == '3':
            # Delete Product
            product_id = input("Enter product ID to delete: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.delete_product(product_id)

        elif choice == '4':
            # View Product
            product_id = input("Enter product ID to view: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.view_product(product_id)

        elif choice == '5':
            # View All Products
            inventory.view_all_products()

        elif choice == '6':
            # Exit
            print("Exiting the system...")
            break

        else:
            print("Invalid choice, please try again.")

if __name__ == "__main__":
    main()



Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 4
Enter product ID to view: 9630868580
Product with ID 9630868580 not found.

Inventory Management System
1. Add Product
2. Update Product
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 6
Exiting the system...


#**5** rewrite the code as, during the operating any function, valid the exit function and redirect the main menu

In [11]:
import datetime
import re

class Product:
    def __init__(self, product_id, name, serial_number, category, price):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = datetime.datetime.now()

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def update_product_name(self, product_id):
        """Update only the product name."""
        if product_id in self.products:
            product = self.products[product_id]
            print(f"Current name: {product.name}")
            new_name = input("Enter new name for the product (type 'exit' to return to the main menu): ").strip()

            if new_name.lower() == 'exit':
                return  # Exiting the function and returning to main menu

            if len(new_name) >= 3:
                product.name = new_name
                print(f"Product name updated to: {new_name}")
            else:
                print("Error: Product name must be at least 3 characters.")
        else:
            print(f"Product with ID {product_id} not found.")

        return  # After updating the name, return to the main menu (no further updates)

    def delete_product(self, product_id):
        if product_id in self.products:
            del self.products[product_id]
            print(f"Product ID {product_id} deleted successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def view_product(self, product_id):
        if product_id in self.products:
            print(self.products[product_id])
        else:
            print(f"Product with ID {product_id} not found.")

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

    def validate_product_id(self, product_id):
        """Validate that product ID is a 10-digit integer."""
        return isinstance(product_id, int) and len(str(product_id)) == 10

    def get_valid_input(self, prompt, validation_func, error_message):
        """Helper function to get valid input from the user with custom validation."""
        while True:
            user_input = input(prompt)
            if user_input.lower() == 'exit':
                return None  # Exit the input and return to main menu

            if validation_func(user_input):
                return user_input
            else:
                print(error_message)

def main():
    inventory = Inventory()

    # Menu for user interaction
    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Update Product Name")
        print("3. Delete Product")
        print("4. View Product")
        print("5. View All Products")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            # Add Product with error handling
            print("\n--- Add Product ---")

            # Get valid product ID
            product_id = inventory.get_valid_input(
                "Enter product ID (exactly 10 digits, or 'exit' to go back): ",
                lambda x: x.isdigit() and len(x) == 10,
                "Error: Product ID must be exactly 10 digits."
            )
            if product_id is None:
                continue
            product_id = int(product_id)

            # Get valid product name
            name = inventory.get_valid_input(
                "Enter product name (at least 3 characters, or 'exit' to go back): ",
                lambda x: len(x) >= 3,
                "Error: Product name must be at least 3 characters."
            )
            if name is None:
                continue

            # Get valid serial number
            serial_number = inventory.get_valid_input(
                "Enter serial number (letters followed by numbers, or 'exit' to go back): ",
                lambda x: bool(re.match(r'^[A-Za-z]+\d+$', x)),
                "Error: Serial number must be a combination of letters and numbers (e.g., ABC1234567)."
            )
            if serial_number is None:
                continue

            category = input("Enter product category: ")
            try:
                price = float(input("Enter product price: "))
            except ValueError:
                print("Error: Price must be a valid number.")
                continue

            product = Product(product_id, name, serial_number, category, price)
            inventory.add_product(product)

        elif choice == '2':
            # Update Product Name
            product_id = input("Enter product ID to update: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)

            inventory.update_product_name(product_id)

        elif choice == '3':
            # Delete Product
            product_id = input("Enter product ID to delete: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.delete_product(product_id)

        elif choice == '4':
            # View Product
            product_id = input("Enter product ID to view: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.view_product(product_id)

        elif choice == '5':
            # View All Products
            inventory.view_all_products()

        elif choice == '6':
            # Exit
            print("Exiting the system...")
            break

        else:
            print("Invalid choice, please try again.")

if __name__ == "__main__":
    main()



Inventory Management System
1. Add Product
2. Update Product Name
3. Delete Product
4. View Product
5. View All Products
6. Exit
Enter your choice: 5
No products in inventory.

Inventory Management System
1. Add Product
2. Update Product Name
3. Delete Product
4. View Product
5. View All Products
6. Exit


KeyboardInterrupt: Interrupted by user

#5 write the code for inserting 50 random values



In [10]:
import random
import string
import datetime
import csv

class Product:
    def __init__(self, product_id, name, serial_number, category, price):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = datetime.datetime.now()

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

    def to_dict(self):
        """Convert Product instance to dictionary for CSV saving."""
        return {
            'product_id': self.product_id,
            'name': self.name,
            'serial_number': self.serial_number,
            'category': self.category,
            'price': self.price,
            'date_added': self.date_added.strftime('%Y-%m-%d %H:%M:%S')
        }

class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def save_to_csv(self, filename="inventory.csv"):
        """Save inventory data to CSV file."""
        with open(filename, 'w', newline='') as csvfile:
            fieldnames = ['product_id', 'name', 'serial_number', 'category', 'price', 'date_added']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

            writer.writeheader()
            for product in self.products.values():
                writer.writerow(product.to_dict())

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

# Function to generate random product details
def generate_random_product():
    product_id = random.randint(1000000000, 9999999999)  # Random 10-digit product ID
    name = ''.join(random.choices(string.ascii_uppercase + string.digits, k=random.randint(3, 8)))  # Random product name (3-8 chars)
    serial_number = ''.join(random.choices(string.ascii_uppercase, k=3)) + str(random.randint(100000, 999999))  # Serial number like ABC1234567
    category = random.choice(["Electronics", "Furniture", "Clothing", "Toys", "Groceries", "Beauty"])
    price = round(random.uniform(10.0, 500.0), 2)  # Random price between 10 and 500

    return Product(product_id, name, serial_number, category, price)

# Function to add multiple random products to the inventory
def add_sample_products(inventory, count=50):
    """Add a specified number of random products to the inventory."""
    for _ in range(count):
        product = generate_random_product()
        inventory.add_product(product)

def main():
    # Initialize the inventory system
    inventory = Inventory()

    # Add 50 sample products to the inventory
    add_sample_products(inventory, 50)

    # Save the products to a CSV file
    inventory.save_to_csv("inventory.csv")

    # View all products after insertion
    print("\nAll Products in Inventory:")
    inventory.view_all_products()

if __name__ == "__main__":
    main()


Product P850 added successfully.
Product HOUQDMZ added successfully.
Product A1G added successfully.
Product FIJIYP78 added successfully.
Product 4YDB added successfully.
Product RBNNK added successfully.
Product LC6FT added successfully.
Product FB6 added successfully.
Product 7UO2 added successfully.
Product HCY added successfully.
Product 7VV added successfully.
Product 1Z2F added successfully.
Product OYUKI added successfully.
Product ZWR added successfully.
Product 4XF added successfully.
Product EARNB added successfully.
Product I0H69H0 added successfully.
Product B95L added successfully.
Product E2JN4R3 added successfully.
Product 4Q1 added successfully.
Product WLDWC5RO added successfully.
Product 8RIF3L added successfully.
Product 9ZAD60UR added successfully.
Product VMGCE2Q added successfully.
Product PZ5VT68N added successfully.
Product V6RDQQL added successfully.
Product 4ZBDF added successfully.
Product UCFG added successfully.
Product I25H added successfully.
Product XAMU

In [ ]:
import datetime
import re
import csv

class Product:
    def __init__(self, product_id, name, serial_number, category, price, date_added=None):
        self.product_id = product_id
        self.name = name
        self.serial_number = serial_number
        self.category = category
        self.price = price
        self.date_added = date_added or datetime.datetime.now()  # Ensure it's a datetime object

    def __str__(self):
        return f"ID: {self.product_id}, Name: {self.name}, Serial: {self.serial_number}, Category: {self.category}, Price: ${self.price:.2f}, Added on: {self.date_added.strftime('%Y-%m-%d %H:%M:%S')}"

    @classmethod
    def from_dict(cls, data):
        """Convert dictionary to Product instance, ensuring date_added is a datetime object."""
        date_added = datetime.datetime.strptime(data['date_added'], '%Y-%m-%d %H:%M:%S')
        return cls(
            product_id=data['product_id'],
            name=data['name'],
            serial_number=data['serial_number'],
            category=data['category'],
            price=float(data['price']),  # Ensure price is a float
            date_added=date_added
        )


class Inventory:
    def __init__(self):
        self.products = {}

    def add_product(self, product):
        if product.product_id in self.products:
            print(f"Product with ID {product.product_id} already exists.")
        else:
            self.products[product.product_id] = product
            print(f"Product {product.name} added successfully.")

    def save_to_csv(self, filename="inventory.csv"):
        """Save inventory data to CSV file."""
        with open(filename, 'w', newline='') as csvfile:
            fieldnames = ['product_id', 'name', 'serial_number', 'category', 'price', 'date_added']
            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)

            writer.writeheader()
            for product in self.products.values():
                writer.writerow({
                    'product_id': product.product_id,
                    'name': product.name,
                    'serial_number': product.serial_number,
                    'category': product.category,
                    'price': product.price,
                    'date_added': product.date_added.strftime('%Y-%m-%d %H:%M:%S')
                })

    def load_from_csv(self, filename="inventory.csv"):
        """Load products from a CSV file into the inventory."""
        try:
            with open(filename, 'r') as csvfile:
                reader = csv.DictReader(csvfile)
                for row in reader:
                    product = Product.from_dict(row)
                    self.add_product(product)
        except FileNotFoundError:
            print("No existing inventory found, starting with an empty inventory.")

    def update_product(self):
        """Update product attributes (name, price, category, etc.) by selecting from available products."""
        if not self.products:
            print("No products in inventory to update.")
            return

        # Display a list of all products with index
        print("\n--- Select a Product to Update ---")
        for idx, product in enumerate(self.products.values(), start=1):
            print(f"{idx}. {product.name} (ID: {product.product_id})")

        # Get the user's choice
        try:
            choice = int(input(f"Enter the number of the product to update (1-{len(self.products)}), or '0' to cancel: "))
            if choice == 0:
                print("Canceling update.")
                return

            # Ensure the choice is valid
            if choice < 1 or choice > len(self.products):
                print("Invalid choice. Please try again.")
                return

            # Select the product based on the choice
            selected_product = list(self.products.values())[choice - 1]
            print(f"\nYou selected: {selected_product.name} (ID: {selected_product.product_id})")

            # Display the update options
            print("\n--- Choose an option to update ---")
            print("1. Update Product Name")
            print("2. Update Product ID (Cannot change once set!)")
            print("3. Update Price")
            print("4. Update Category")
            print("5. Update Date Added (Cannot change once set!)")
            update_choice = input("Enter your choice (1-5), or '0' to cancel: ")

            if update_choice == '0':
                print("Canceling update.")
                return

            if update_choice == '1':
                new_name = input(f"Enter new name for the product (current name: {selected_product.name}): ").strip()
                if len(new_name) >= 3:
                    selected_product.name = new_name
                    print(f"Product name updated to: {new_name}")
                else:
                    print("Error: Product name must be at least 3 characters.")

            elif update_choice == '2':
                print("Product ID cannot be changed once set.")
                return

            elif update_choice == '3':
                try:
                    new_price = float(input(f"Enter new price for the product (current price: {selected_product.price}): "))
                    if new_price > 0:
                        selected_product.price = new_price
                        print(f"Product price updated to: ${new_price:.2f}")
                    else:
                        print("Error: Price must be a positive number.")
                except ValueError:
                    print("Error: Invalid price.")

            elif update_choice == '4':
                new_category = input(f"Enter new category for the product (current category: {selected_product.category}): ").strip()
                if new_category:
                    selected_product.category = new_category
                    print(f"Product category updated to: {new_category}")
                else:
                    print("Error: Category cannot be empty.")

            elif update_choice == '5':
                print("Product date added cannot be changed once set.")
                return

            else:
                print("Invalid choice. Please try again.")

        except ValueError:
            print("Invalid input. Please enter a number.")

    def delete_product(self, product_id):
        if product_id in self.products:
            del self.products[product_id]
            print(f"Product ID {product_id} deleted successfully.")
        else:
            print(f"Product with ID {product_id} not found.")

    def view_product(self, product_id):
        if product_id in self.products:
            print(self.products[product_id])
        else:
            print(f"Product with ID {product_id} not found.")

    def view_all_products(self):
        if not self.products:
            print("No products in inventory.")
        else:
            for product in self.products.values():
                print(product)

    def get_valid_input(self, prompt, validation_func, error_message):
        """Helper function to get valid input from the user with custom validation."""
        while True:
            user_input = input(prompt)
            if user_input.lower() == 'exit':
                return None  # Exit the input and return to main menu

            if validation_func(user_input):
                return user_input
            else:
                print(error_message)


def main():
    inventory = Inventory()

    # Load products from CSV (if any)
    inventory.load_from_csv()

    # Menu for user interaction
    while True:
        print("\nInventory Management System")
        print("1. Add Product")
        print("2. Update Product")
        print("3. Delete Product")
        print("4. View Product")
        print("5. View All Products")
        print("6. Exit")

        choice = input("Enter your choice: ")

        if choice == '1':
            # Add Product with error handling
            print("\n--- Add Product ---")

            # Get valid product ID
            product_id = inventory.get_valid_input(
                "Enter product ID (exactly 10 digits, or 'exit' to go back): ",
                lambda x: x.isdigit() and len(x) == 10,
                "Error: Product ID must be exactly 10 digits."
            )
            if product_id is None:
                continue
            product_id = int(product_id)

            # Get valid product name
            name = inventory.get_valid_input(
                "Enter product name (at least 3 characters, or 'exit' to go back): ",
                lambda x: len(x) >= 3,
                "Error: Product name must be at least 3 characters."
            )
            if name is None:
                continue

            # Get valid serial number
            serial_number = inventory.get_valid_input(
                "Enter serial number (letters followed by numbers, or 'exit' to go back): ",
                lambda x: bool(re.match(r'^[A-Za-z]+\d+$', x)),
                "Error: Serial number must be a combination of letters and numbers (e.g., ABC1234567)."
            )
            if serial_number is None:
                continue

            category = input("Enter product category: ")
            try:
                price = float(input("Enter product price: "))
            except ValueError:
                print("Error: Price must be a valid number.")
                continue

            product = Product(product_id, name, serial_number, category, price)
            inventory.add_product(product)

        elif choice == '2':
            # Update Product
            inventory.update_product()

        elif choice == '3':
            # Delete Product
            product_id = input("Enter product ID to delete: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.delete_product(product_id)

        elif choice == '4':
            # View Product
            product_id = input("Enter product ID to view: ")
            if not product_id.isdigit():
                print("Error: Product ID should be numeric.")
                continue
            product_id = int(product_id)
            inventory.view_product(product_id)

        elif choice == '5':
            # View All Products
            inventory.view_all_products()

        elif choice == '6':
            # Save to CSV and Exit
            inventory.save_to_csv()
            print("Exiting the system...")
            break

        else:
            print("Invalid choice, please try again.")

if __name__ == "__main__":
    main()


Product P850 added successfully.
Product HOUQDMZ added successfully.
Product A1G added successfully.
Product FIJIYP78 added successfully.
Product 4YDB added successfully.
Product RBNNK added successfully.
Product LC6FT added successfully.
Product FB6 added successfully.
Product 7UO2 added successfully.
Product HCY added successfully.
Product 7VV added successfully.
Product 1Z2F added successfully.
Product OYUKI added successfully.
Product ZWR added successfully.
Product 4XF added successfully.
Product EARNB added successfully.
Product I0H69H0 added successfully.
Product B95L added successfully.
Product E2JN4R3 added successfully.
Product 4Q1 added successfully.
Product WLDWC5RO added successfully.
Product 8RIF3L added successfully.
Product 9ZAD60UR added successfully.
Product VMGCE2Q added successfully.
Product PZ5VT68N added successfully.
Product V6RDQQL added successfully.
Product 4ZBDF added successfully.
Product UCFG added successfully.
Product I25H added successfully.
Product XAMU

# Question 7 (20 points)

Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

In [ ]:
# Your code here


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below